bitmark-references/rpc-communication/rpc-communication.md

```diff
@@ -0,0 +1,117 @@
+# RPC
+
+## Introduction
+
+RPC is the abbreviation of remote procedure call, which is an inter-process communication technique used for client-server based applications. It is also known as subroutine call or function call.
+
+# Flow
+
+## Initialization
+
+When `bitmarkd` starts, it initialises RPC settings including maximum connection count, connection bandwidth, communication encryption key, binds RPC on specific port, listen RPC port.
+
+## Processing Messages
+
+Different RPC command is processed by differnet function, following sections describes each command with examples. Each message is in json format, with one key `method` denotes which remote method being executed, and the other key `params` denotes what arguments are passed.
+
+In order to test example, `bitmarkd` should be started in advance, default RPC port is on `2130`. `openssl` should be installed in advance to run communication.
+
+All methods are named in the format of `file_name.method_name`, for example, `Assets.Get` stands for method `Get` in file `assets.go`.
+
+Available commands:
+
+1. Assets.Get - retrieve assets information
+
+It is necessary to provide array of asset fingerprints to query asset information fingerprints).
+
+Example shell command:
+
+```shell
+(echo '{"method":"Assets.Get","params":[{"fingerprints": ["01c7499e5d27880da75a3747340aa1cb2f11fa023f7aa1eb10acbf28e447aefd366759092de7a31fdfe89fcb88ecc3c90c0e067484184f41a8e3043e8aa4732f00", "01f9c2eed799128d331c0e60b07d99bffa299b100eea7bb4a410d8f2ee2a04218623f2dbd1a53f0fe08f9cda131ecff213889cbd2cf5c8e53100581ff00f6270c1"]}]}'; sleep 0.5; echo Q) | openssl s_client -connect 127.0.0.1:2130 -quiet
+```
+
+1. Bitmark.Provenance - retrieve transfer history of a bitmark
+
+In order to get provenance (asset transfer history), two arguments need to be provided, one is number of history to get (count), the other is transaction ID (txID).
+
+Example shell command:
+
+```shell
+(echo '{"method":"Bitmark.Provenance","params":[{"count":20,"txId":"2dc8770718b01f0205ad991bfb4c052f02677cff60e65d596e890cb6ed82c861"}]}'; sleep 0.5; echo Q) | openssl s_client -connect 127.0.0.1:2130 -quiet
+```
+
+1. Bitmark.Transfer - transfer a bitmark
+
+To transfer a bitmark, some data is needed: previous transaction hash (link), next owner account (owner), previous hex string of owner (signature), next hex string of owner (countersignature).
+
+Example shell command:
+
+```shell
+(echo '{"id":"1","method":"Bitmark.Transfer","params":[{"link":"1bebd06c8ecb8b11ea93e93c9d38b7f6d7dfdf015530819015172cf51c7f33f7", "owner": "eZpG6Wi9SQvpDatEP7QGrx6nvzwd6s6R8DgMKgDbDY1R5bjzb9", "signature": "a3e456a31a4a64962a32bcbf6549d14134deeb5d87285a04c648355eb9e59d938f8ab440d2b50c781baf2c1a5a2112c2167301bb128c8f850a9d54f3b27c5a08"}]}'; sleep 0.5; echo Q) | openssl s_client -connect 127.0.0.1:2130 -quiet
+```
+
```

Comment:
We need to explain "id", since the intro just said we'd be using "method" and "params"

Reply:
remove it